### PR TITLE
Modular Assistive Driving System (MADS)

### PR DESCRIPTION
**Pull request by sunnyhaibin**
Supersedes https://github.com/sunnypilot/panda/pull/52

### CODE REVIEW

**Comment by sunnyhaibin** [comment]
@devtekve ready for review. Fixed disengage reasons not being emitted as expected during control exits.
image

File: board/drivers/gpio.h [hunk 1]
```diff
@@ -58,7 +58,7 @@ void set_gpio_alternate(GPIO_TypeDef *GPIO, unsigned int pin, unsigned int mode)
 void set_gpio_pullup(GPIO_TypeDef *GPIO, unsigned int pin, unsigned int mode) {
   ENTER_CRITICAL();
   uint32_t tmp = GPIO->PUPDR;
-  tmp &= ~(3U << (pin * 2U));
+  tmp &= ~((uint32_t)3U << ((uint32_t)(pin * 2U)));
```
Comment (devtekve) [inline board/drivers/gpio.h:61]: Sorry for this. MISRA decided to validate this path when we have mads_exit_controls() due to LAG. But also on the braking... Like, I dont know why but commenting either makes MISRA happy and doesn't complain here. Adding them somehow makes MISRA reconsider it's life choices and comes and checks something unrelated....

**Review by devtekve — APPROVED** [review]

File: board/drivers/gpio.h (hunk 1 above)
Comment (sunnyhaibin) [reply board/drivers/gpio.h]: Resolved and reverted in https://github.com/sunnypilot/panda/commit/6dab7bbb4787aa13a17242bb4c445a18c67d81c7.

File: board/sunnypilot/safety_mads.h [hunk 2]
```diff
@@ -195,5 +199,7 @@ inline void mads_state_update(const bool op_vehicle_moving, const bool op_acc_ma
   m_update_binary_state(&m_mads_state.braking);
   m_update_button_state(&m_mads_state.mads_button);
 
-  m_update_control_state();
+  const bool mads_allowed = m_update_control_state();
+
+  m_publish_state(mads_allowed);
```
Comment (devtekve) [inline board/sunnypilot/safety_mads.h:204]: We are not publishing the state anywhere, we are updating it

File: board/sunnypilot/safety_mads.h [hunk 3]
```diff
@@ -111,28 +111,32 @@ static void m_update_control_state(void) {
...
+      allowed = false;
     } else {
-      // Do nothing
     }
   }
 
+  return allowed;
+}
+
+static void m_publish_state(bool allowed) {
```
Comment (devtekve) [inline board/sunnypilot/safety_mads.h:137]: If you weren't allowed, why would you call this method?  and if you are allowed, do you really need to pass it as argument?

Comment (devtekve) [reply board/sunnypilot/safety_mads.h:137]: And following that line of why do you need an argument on your method if it makes no sense, what about getting the allowed from `m_update_control_state()` directly inside of this method?

**Review by devtekve — DISMISSED** [review]

File: tests/safety/test_honda.py [hunk 4]
```diff
@@ -361,7 +404,7 @@ def setUp(self):
     self.safety.init_tests()
 
   def _acc_state_msg(self, main_on):
-    values = {"MAIN_ON": main_on, "COUNTER": self.cnt_acc_state % 4}
+    values = {"MAIN_ON": 1 if main_on else 0, "COUNTER": self.cnt_acc_state % 4}
```
Comment (sunnyhaibin) [inline tests/safety/test_honda.py:407]:
```suggestion
    values = {"MAIN_ON": main_on, "COUNTER": self.cnt_acc_state % 4}
```

**Pull request merged**